### ISSUE TITLE
Improve synchronization syntax

### ISSUE THREAD

**Issue by araex** [araex/BannerlordCoop#8]
Adding methods & fields to be synchronized via the persistence library is tedious and error prone due to the necessity of naming all parameters of a function in patch explicitly. Failing to notice a change in the API (even just a renaming of a parameter of a function) will lead to runtime errors when initializing harmony.

Instead, provide an easy to use syntax without redundancies and as few references to the games DLLs as possible.

The branch `dev_RPC` contains a prototype using a fluent interface for synchronized method calls:
```
[Patch]
public static class MyPatch
{
private static MethodPatcher m_Patcher = new MethodPatcher(typeof(Foo))
   .Synchronize(nameof(Foo.Bar))
   .Synchronize(nameof(Foo.Baz));
}
```
This will patch a dynamically generated prefix containing all the original parameters. Internally, the call is dispatched to a generic `Call(object instance, object[] args`. `RPCSyncHandlers` will pick up the `[Patch]` attribute and bind the persistence library to the auto generated prefixes. The arguments are not fully serialized, instead they are transfered as references to existing game state, for example a `MBGUID` or a `RailEntity`. The persistence library is able to resolve those references before calling the original function on the receiving end.

Effectively, the `MyPatch` above is all that is necessary to fully synchronize any function call between all clients. However, every individual argument type has to be supported explicitly for encoding and decoding in the persistence library.

Im not entirely sure if the fluent syntax is good for this purpose, it might get confusing.

There's still a couple of issues with the prototype. I would also like to use a similar syntax for the field synchronization `[SyncField]`. Preferably eliminating manual handling of `SyncField` and `SyncMethod` entirely.

**Comment by araex** [comment]
The system would be very useful to synchronize something like `TaleWorlds.CampaignSystem.Actions.GiveGoldAction`. Currently one would have to write 9 individual patches, each containing explicity named parameter and methods. Most of the patches would probably not change any game logic but rather just synchronize the call. That is a lot of redundant code that might break with every update of the game.

The new prototype can auto generate those default patches for the whole class, without naming anything explicitly except for the classes type. This improves maintainability in regards to new game versions drastically. A considerable drawback however is the added complexity in the backend and what might appear like "magic" to someone not familiar with the code base.

**Comment by araex** [comment]
Merged with 1d37f8d. Synchronizing an action or field consist of two separate phases:
1. Define the patch
2. Define the synchronization mechanism

For patching the helper classes `MethodPatch` and `PropertyPatch` can be used. They dynamically generate matching methods and do the patching automatically. 

Synchronization is added trough `MethodAccess.SetHandler`. The `MethodAccess` is obtained using the `MethodPatch.Methods`, `PropertyPatch.Setters` and `PropertyPatch.Getters`. This process can be automated for synchronization trough the persistence library, see `RpcSyncHandlers.Register`.

Final syntax to patch an action:
```
public static class RecruitPatch
{
    private static readonly MethodPatch Patch =
        new MethodPatch(typeof(TaleWorlds.CampaignSystem.RecruitAction))
            .InterceptAll(BindingFlags.Static | BindingFlags.Public | BindingFlags.DeclaredOnly);

    [PatchInitializer]
    public static void Init()
    {
        CoopClient.Instance.OnPersistenceInitialized += persistence =>
            persistence.RpcSyncHandlers.Register(Patch.Methods);

        foreach (MethodAccess method in Patch.Methods)
        {
            method.Condition = Coop.DoSync;
        }
    }
}
```
